lib/srv/forward/sshserver.go

```diff
@@ -862,28 +874,128 @@ func (s *Server) rejectChannel(chans <-chan ssh.NewChannel, errMessage string) {
+			s.log.WithError(err).Warnf("Failed to cancel tcpip forward request")
+			if err := req.Reply(false, nil); err != nil {
+				s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			}
+			return
+		}
+	default:
+		ok, payload, err := s.remoteClient.SendRequest(ctx, req.Type, req.WantReply, req.Payload)
+		if err != nil {
+			s.log.Warnf("Failed to forward global request %v: %v", req.Type, err)
+			return
+		}
+		if req.WantReply {
+			if err := req.Reply(ok, payload); err != nil {
+				s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			}
+		}
 	}
+}
 
-	ok, payload, err := s.remoteClient.SendRequest(ctx, req.Type, req.WantReply, req.Payload)
+// handleTCPIPForwardRequest handles remote port forwarding requests.
+func (s *Server) handleTCPIPForwardRequest(ctx context.Context, r *ssh.Request) error {
+	req, err := sshutils.ParseTCPIPForwardRequest(r.Payload)
 	if err != nil {
-		s.log.Warnf("Failed to forward global request %v: %v", req.Type, err)
-		return
+		return trace.Wrap(err)
 	}
 
-	if req.WantReply {
-		err = req.Reply(ok, payload)
+	// On forward server in "tcpip-forward" requests from SessionJoinPrincipal
+	//  should be rejected, otherwise it's possible to use the
+	// "-teleport-internal-join" user to bypass RBAC.
+	if s.identityContext.Login == teleport.SSHSessionJoinPrincipal {
+		s.log.Error("Request rejected, tcpip-forward with SessionJoinPrincipal in forward node must be blocked")
+		err := trace.AccessDenied("attempted tcpip-forward request in join-only mode")
+		if r.WantReply {
+			if replyErr := r.Reply(false, []byte(utils.FormatErrorWithNewline(err))); replyErr != nil {
+				s.log.Errorf("sending error reply to SSH global request: %v", replyErr)
+			}
+		}
+		return err
+	}
+
+	ctx, scx, err := srv.NewServerContext(ctx, s.connectionContext, s, s.identityContext)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	listenAddr := sshutils.JoinHostPort(req.Addr, req.Port)
+	scx.SrcAddr = listenAddr
+	scx.DstAddr = scx.NetConn.RemoteAddr().String()
+
+	if err := s.authHandlers.CheckPortForward(scx.DstAddr, scx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	listener, err := s.remoteClient.Listen("tcp", listenAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	scx.AddCloser(listener)
+	// Set the src addr again since it may have been updated with a new port.
+	scx.SrcAddr = listener.Addr().String()
+	s.remoteForwardingMap.Store(scx.SrcAddr, scx)
+
+	if err := sshutils.StartRemoteListener(ctx, scx.ConnectionContext.ServerConn, scx.SrcAddr, scx.DstAddr, listener); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Report addr back to the client.
+	if r.WantReply {
+		_, port, err := sshutils.SplitHostPort(listener.Addr().String())
 		if err != nil {
-			s.log.Warnf("Failed to reply to global request: %v: %v", req.Type, err)
+			return trace.Wrap(err)
+		}
+
+		var buf [4]byte
+		binary.BigEndian.PutUint32(buf[:], port)
```

Comment:
`ssh.Marshal` on a `struct { Port uint32 }` would make it more obvious that this is just part of the SSH spec.